File: rez-go-sdk/rez/services.go
/*
REZ Go SDK - Invoice, Contracts, Legal, Twin, Ranking, GraphQL Services
*/
package rez

import (
	"context"
	"encoding/json"
	"fmt"
)

// ============================================================================
// Invoice Service
// ============================================================================

// InvoiceService handles invoice operations
type InvoiceService struct {
	client *Client
}

// CreateInvoice creates a new invoice
func (s *InvoiceService) Create(ctx context.Context, req InvoiceCreateRequest) (*InvoiceResponse, error) {
	data, status, err := s.client.doPost(ctx, "/api/invoice/create", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp InvoiceResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// GetInvoice retrieves an invoice by ID
func (s *InvoiceService) Get(ctx context.Context, invoiceID string) (*InvoiceResponse, error) {
	data, status, err := s.client.doGet(ctx, "/api/invoice/"+invoiceID, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp InvoiceResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// List retrieves all invoices
func (s *InvoiceService) List(ctx context.Context, skip, limit int, status string) (*InvoiceListResponse, error) {
	params := map[string]string{
		"skip":  fmt.Sprintf("%d", skip),
		"limit": fmt.Sprintf("%d", limit),
	}
	if status != "" {
		params["status"] = status
	}

	data, status, err := s.client.doGet(ctx, "/api/invoice/list", params)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp InvoiceListResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// Validate validates an invoice
func (s *InvoiceService) Validate(ctx context.Context, invoiceID string) (map[string]interface{}, error) {
	data, status, err := s.client.doPost(ctx, "/api/invoice/validate/"+invoiceID, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// ============================================================================
// Contracts Service
// ============================================================================

// ContractsService handles contract operations
type ContractsService struct {
	client *Client
}

// Generate generates a new contract
func (s *ContractsService) Generate(ctx context.Context, req ContractGenerateRequest) (*ContractResponse, error) {
	data, status, err := s.client.doPost(ctx, "/api/contracts/generate", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp ContractResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// Get retrieves a contract by ID
func (s *ContractsService) Get(ctx context.Context, contractID string) (*ContractResponse, error) {
	data, status, err := s.client.doGet(ctx, "/api/contracts/"+contractID, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp ContractResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// Analyze analyzes a contract
func (s *ContractsService) Analyze(ctx context.Context, contractID string) (map[string]interface{}, error) {
	data, status, err := s.client.doPost(ctx, "/api/contracts/analyze/"+contractID, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// ============================================================================
// Legal Service
// ============================================================================

// LegalService handles legal operations
type LegalService struct {
	client *Client
}

// Research performs legal research
func (s *LegalService) Research(ctx context.Context, query, jurisdiction string) (map[string]interface{}, error) {
	path := "/api/legal/research?query=" + query
	if jurisdiction != "" {
		path += "&jurisdiction=" + jurisdiction
	}

	data, status, err := s.client.doGet(ctx, path, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// AnalyzeDocument analyzes a legal document
func (s *LegalService) AnalyzeDocument(ctx context.Context, documentText string) (map[string]interface{}, error) {
	req := map[string]string{"document_text": documentText}
	data, status, err := s.client.doPost(ctx, "/api/legal/analyze", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// CheckCompliance checks compliance
func (s *LegalService) CheckCompliance(ctx context.Context, requirements []string, context_ map[string]interface{}) (map[string]interface{}, error) {
	req := map[string]interface{}{
		"requirements": requirements,
		"context":      context_,
	}
	data, status, err := s.client.doPost(ctx, "/api/legal/compliance", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// ============================================================================
// Twin Service
// ============================================================================

// TwinService handles digital twin operations
type TwinService struct {
	client *Client
}

// Create creates a new digital twin
func (s *TwinService) Create(ctx context.Context, req TwinCreateRequest) (*TwinResponse, error) {
	data, status, err := s.client.doPost(ctx, "/api/twin/create", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp TwinResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// Get retrieves a twin by ID
func (s *TwinService) Get(ctx context.Context, twinID string) (*TwinResponse, error) {
	data, status, err := s.client.doGet(ctx, "/api/twin/"+twinID, nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp TwinResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// GetState retrieves twin state
func (s *TwinService) GetState(ctx context.Context, twinID string) (map[string]interface{}, error) {
	data, status, err := s.client.doGet(ctx, "/api/twin/"+twinID+"/state", nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// UpdateState updates twin state
func (s *TwinService) UpdateState(ctx context.Context, twinID string, state map[string]interface{}) (map[string]interface{}, error) {
	req := TwinStateUpdateRequest{State: state}
	data, status, err := s.client.doPost(ctx, "/api/twin/"+twinID+"/state", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// Sync synchronizes twin
func (s *TwinService) Sync(ctx context.Context, twinID string) (map[string]interface{}, error) {
	data, status, err := s.client.doPost(ctx, "/api/twin/"+twinID+"/sync", nil)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp, nil
}

// ============================================================================
// Ranking Service
// ============================================================================

// RankingService handles ranking operations
type RankingService struct {
	client *Client
}

// ScoreEntities scores and ranks entities
func (s *RankingService) Score(ctx context.Context, req RankingScoreRequest) (*RankingScoreResponse, error) {
	data, status, err := s.client.doPost(ctx, "/api/ranking/score", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp RankingScoreResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

// GetTopK retrieves top K ranked entities
func (s *RankingService) GetTopK(ctx context.Context, entities []map[string]interface{}, k int, algorithm string) ([]map[string]interface{}, error) {
	req := map[string]interface{}{
		"entities":  entities,
		"k":        k,
		"algorithm": algorithm,
	}
	data, status, err := s.client.doPost(ctx, "/api/ranking/top-k", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp struct {
		Entities []map[string]interface{} `json:"entities"`
	}
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return resp.Entities, nil
}

// ============================================================================
// GraphQL Service
// ============================================================================

// GraphQLService handles GraphQL operations
type GraphQLService struct {
	client *Client
}

// Execute executes a GraphQL query
func (s *GraphQLService) Execute(ctx context.Context, req GraphQLRequest) (*GraphQLResponse, error) {
	data, status, err := s.client.doPost(ctx, "/graphql", req)
	if err != nil {
		return nil, err
	}

	if status >= 400 {
		return nil, fmt.Errorf("API error: %d", status)
	}

	var resp GraphQLResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &resp, nil
}

File: rez-java-sdk/src/main/java/io/rez/sdk/Services.java
package io.rez.sdk;

import java.util.*;

/**
 * REZ Java SDK - Agent Service
 */
public class AgentService {
    private final REZClient client;

    public AgentService(REZClient client) {
        this.client = client;
    }

    public Models.AgentList list(int skip, int limit) throws REZException {
        Map<String, String> params = new HashMap<>();
        params.put("skip", String.valueOf(skip));
        params.put("limit", String.valueOf(limit));
        Models.AgentListResponse response = client.get("/api/agents", params, Models.AgentListResponse.class);
        return new Models.AgentList(response.agents, response.total, response.skip, response.limit);
    }

    public Models.AgentList list() throws REZException {
        return list(0, 20);
    }

    public Models.Agent get(String agentId) throws REZException {
        return client.get("/api/agents/" + agentId, Models.Agent.class);
    }

    public Models.Agent create(Models.AgentCreateRequest request) throws REZException {
        return client.post("/api/agents", request, Models.Agent.class);
    }

    public Models.Agent update(String agentId, Models.AgentUpdateRequest request) throws REZException {
        return client.put("/api/agents/" + agentId, request, Models.Agent.class);
    }

    public void delete(String agentId) throws REZException {
        client.delete("/api/agents/" + agentId);
    }

    public Models.HealthResponse healthCheck() throws REZException {
        return client.get("/health", Models.HealthResponse.class);
    }
}

/**
 * REZ Java SDK - AutoML Service
 */
class AutoMLService {
    private final REZClient client;

    AutoMLService(REZClient client) {
        this.client = client;
    }

    public List<Models.AutoMLModel> listModels(int skip, int limit) throws REZException {
        Map<String, String> params = new HashMap<>();
        params.put("skip", String.valueOf(skip));
        params.put("limit", String.valueOf(limit));
        Models.ModelsResponse response = client.get("/api/automl/models", params, Models.ModelsResponse.class);
        return response.models != null ? response.models : Collections.emptyList();
    }

    public Models.AutoMLModel getModel(String modelId) throws REZException {
        return client.get("/api/automl/models/" + modelId, Models.AutoMLModel.class);
    }

    public Models.AutoMLModel train(Models.AutoMLTrainRequest request) throws REZException {
        return client.post("/api/automl/train", request, Models.AutoMLModel.class);
    }

    public Map<String, Object> predict(String modelId, Models.AutoMLPredictRequest request) throws REZException {
        return client.post("/api/automl/predict/" + modelId, request, Map.class);
    }

    public void deleteModel(String modelId) throws REZException {
        client.delete("/api/automl/models/" + modelId);
    }
}

/**
 * REZ Java SDK - Invoice Service
 */
class InvoiceService {
    private final REZClient client;

    InvoiceService(REZClient client) {
        this.client = client;
    }

    public Models.InvoiceResponse create(Models.InvoiceCreateRequest request) throws REZException {
        return client.post("/api/invoice/create", request, Models.InvoiceResponse.class);
    }

    public Models.InvoiceResponse get(String invoiceId) throws REZException {
        return client.get("/api/invoice/" + invoiceId, Models.InvoiceResponse.class);
    }

    public Models.InvoiceList list(int skip, int limit, String status) throws REZException {
        Map<String, String> params = new HashMap<>();
        params.put("skip", String.valueOf(skip));
        params.put("limit", String.valueOf(limit));
        if (status != null) params.put("status", status);
        Models.InvoiceListResponse response = client.get("/api/invoice/list", params, Models.InvoiceListResponse.class);
        return new Models.InvoiceList(response.invoices, response.total, response.skip, response.limit);
    }

    public Map<String, Object> validate(String invoiceId) throws REZException {
        return client.post("/api/invoice/validate/" + invoiceId, null, Map.class);
    }
}

/**
 * REZ Java SDK - Contracts Service
 */
class ContractsService {
    private final REZClient client;

    ContractsService(REZClient client) {
        this.client = client;
    }

    public Models.ContractResponse generate(Models.ContractGenerateRequest request) throws REZException {
        return client.post("/api/contracts/generate", request, Models.ContractResponse.class);
    }

    public Models.ContractResponse get(String contractId) throws REZException {
        return client.get("/api/contracts/" + contractId, Models.ContractResponse.class);
    }

    public Map<String, Object> analyze(String contractId) throws REZException {
        return client.post("/api/contracts/analyze/" + contractId, null, Map.class);
    }
}

/**
 * REZ Java SDK - Legal Service
 */
class LegalService {
    private final REZClient client;

    LegalService(REZClient client) {
        this.client = client;
    }

    public Map<String, Object> research(String query, String jurisdiction) throws REZException {
        String path = "/api/legal/research?query=" + query;
        if (jurisdiction != null) {
            path += "&jurisdiction=" + jurisdiction;
        }
        return client.get(path, Map.class);
    }

    public Map<String, Object> analyzeDocument(String documentText) throws REZException {
        Map<String, String> request = new HashMap<>();
        request.put("document_text", documentText);
        return client.post("/api/legal/analyze", request, Map.class);
    }

    public Map<String, Object> checkCompliance(List<String> requirements, Map<String, Object> context) throws REZException {
        Map<String, Object> request = new HashMap<>();
        request.put("requirements", requirements);
        request.put("context", context != null ? context : new HashMap<>());
        return client.post("/api/legal/compliance", request, Map.class);
    }
}

/**
 * REZ Java SDK - Twin Service
 */
class TwinService {
    private final REZClient client;

    TwinService(REZClient client) {
        this.client = client;
    }

    public Models.TwinResponse create(Models.TwinCreateRequest request) throws REZException {
        return client.post("/api/twin/create", request, Models.TwinResponse.class);
    }

    public Models.TwinResponse get(String twinId) throws REZException {
        return client.get("/api/twin/" + twinId, Models.TwinResponse.class);
    }

    public Map<String, Object> getState(String twinId) throws REZException {
        return client.get("/api/twin/" + twinId + "/state", Map.class);
    }

    public Map<String, Object> updateState(String twinId, Map<String, Object> state) throws REZException {
        return client.post("/api/twin/" + twinId + "/state", new Models.TwinStateUpdateRequest(state), Map.class);
    }

    public Map<String, Object> sync(String twinId) throws REZException {
        return client.post("/api/twin/" + twinId + "/sync", null, Map.class);
    }
}

/**
 * REZ Java SDK - Ranking Service
 */
class RankingService {
    private final REZClient client;

    RankingService(REZClient client) {
        this.client = client;
    }

    public Models.RankingScoreResponse score(Models.RankingScoreRequest request) throws REZException {
        return client.post("/api/ranking/score", request, Models.RankingScoreResponse.class);
    }

    public List<Map<String, Object>> getTopK(List<Map<String, Object>> entities, int k, String algorithm) throws REZException {
        Map<String, Object> request = new HashMap<>();
        request.put("entities", entities);
        request.put("k", k);
        request.put("algorithm", algorithm);
        Models.TopKResponse response = client.post("/api/ranking/top-k", request, Models.TopKResponse.class);
        return response.entities;
    }
}

/**
 * REZ Java SDK - GraphQL Service
 */
class GraphQLService {
    private final REZClient client;

    GraphQLService(REZClient client) {
        this.client = client;
    }

    public Models.GraphQLResponse execute(String query) throws REZException {
        return client.post("/graphql", new Models.GraphQLRequest(query), Models.GraphQLResponse.class);
    }

    public Models.GraphQLResponse execute(String query, Map<String, Object> variables) throws REZException {
        return client.post("/graphql", new Models.GraphQLRequest(query, variables), Models.GraphQLResponse.class);
    }
}